File: leetcode/108/P108.rb
# 108. Convert Sorted Array to Binary Search Tree
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {Integer[]} nums
# @return {TreeNode}
def sorted_array_to_bst(nums)
	build nums, 0, nums.size - 1
end

def build nums, l, r
	return nil if l > r
	mid = (l + r) / 2
	node = TreeNode.new nums[mid]
	node.left = build nums, l, mid - 1
	node.right = build nums, mid + 1, r
	node
end

File: leetcode/108/src/P108.java
/**
 * Created by qwordy on 2/24/16.
 * 108. Convert Sorted Array to Binary Search Tree
 */

class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;
	TreeNode(int x) { val = x; }
}

class Solution {
	public TreeNode sortedArrayToBST(int[] nums) {
		return build(nums, 0, nums.length - 1);
	}

	private TreeNode build(int[] nums, int l, int r) {
		if (l > r) return null;
		int mid = (l + r) / 2;
		TreeNode node = new TreeNode(nums[mid]);
		node.left = build(nums, l, mid - 1);
		node.right = build(nums, mid + 1, r);
		return node;
	}
}

public class P108 {
}
